### PR TITLE
Nr. 1 - [TEP-0056]: Add Pipelines-in-Pipelines task resolution for `PipelineSpec`.

### PR DESCRIPTION
**Pull request by twoGiants**
> [!IMPORTANT]
> **Pull request number 1.**
> The numbering `Nr. 1 - [TEP-0056]` means the PRs must be merged in order. They build on each other.

# Changes

[[TEP-0056](https://github.com/tektoncd/community/blob/main/teps/0056-pipelines-in-pipelines.md)]: First PR of **Pipelines-in-Pipelines** feature implementation. The `PipelineSpec` functionality is implemented first.

Extend `ResolvedPipelineTask` data structure to store child `PipelineRuns` and the resolved pipeline in `ResolvedPipeline` data structure equal to `ResolvedTask`.

Add `GetPipelineRun` function alias for convenience.

Add condition to `func ResolvePipelineTask` which checks if the given `PipelineTask` is a child (PinP) `Pipeline` and resolves it by getting the actual `Spec`. In our "happy path" case it's right on the `PipelineTask` because we use `PipelineSpec` and don't need to fetch it from anywhere.

A few downstream methods with tests equal to resolving a `TaskRun` were added.

Extend `resolvePipelineState` with getter for child `PipelineRuns` using lister.

Issue #8760, #7166.

Release notes will be added with the last PR which will make this feature functional for users.

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] [pre-commit](https://github.com/tektoncd/pipeline/blob/main/DEVELOPMENT.md#install-tools) Passed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: pkg/reconciler/pipelinerun/resources/pipelinespec.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package resources
+
+import v1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1"
+
+// ResolvedPipeline contains the data that is needed to execute
+// a child (PinP) PipelineRun.
```
Comment (waveywaves) [inline pkg/reconciler/pipelinerun/resources/pipelinespec.go:6]: nit: can we rewrite this as Pipeline in Pipeline (in rest of the places as well) for clarity

**Review by waveywaves — APPROVED** [review]
I believe this PR looks good, just have one suggestion, and based on https://github.com/tektoncd/pipeline/pull/8873#issuecomment-3106938705 we need to cover testing some nil references, apart from that it looks good

/approve

**Comment by waveywaves** [comment]
cc @tektoncd/core-maintainers

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 2]
```diff
@@ -142,10 +146,29 @@ func (t ResolvedPipelineTask) IsCustomTask() bool {
 	return t.CustomTask
 }
 
+// IsChildPipeline returns true if the PipelineTask references a child (PinP) Pipeline.
+func (t ResolvedPipelineTask) IsChildPipeline() bool {
```
Comment (khrm) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:150]: Shouldn't this be `HasChildPipeline` ?

Comment (khrm) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:150]: It's fine. I understand now. I think we should change the comment.  At the moment, this function check whether current task embed a PipelineSpec.

Comment (twoGiants) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:150]: I used the same type of comment as in line 145 above.

```go
// IsCustomTask returns true if the PipelineTask references a Custom Task.
``` 

How would you change the comment? 

The check will be extended once we also check for `PipelineRef`. If one of those two fields is set we have an embedded (i.e. child) pipeline.

**Comment by waveywaves** [comment]
@khrm can you take a look at this one and approve/request changes it if possible?

File: pkg/reconciler/pipelinerun/resources/pipelinespec.go (hunk 1 above)
Comment (twoGiants) [reply pkg/reconciler/pipelinerun/resources/pipelinespec.go:6]: Just saw this comment! Will update everywhere.

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 3]
```diff
@@ -310,15 +359,29 @@ func (t ResolvedPipelineTask) isScheduled() bool {
 	return len(t.TaskRuns) > 0
 }
 
-// haveAnyRunsFailed returns true when any of the taskRuns/customRuns have succeeded condition with status set to false
+// haveAnyRunsFailed returns true when any of the child (PinP) PipelineRuns/TaskRuns/CustomRuns have succeeded condition with status set to false
```
Comment (afrittoli) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:362]: NIT: I referred in the past to pipelines-in-pipelines as pip, which can be confused with the Python tool. 
PinP avoids that confusion, but it also sounds like a not-so-nice word. Not sure what the official acronym should be :P

**Review by afrittoli — COMMENTED** [review]
Thanks for this! 

It looks good, it pretty much follows the patter of TaskRuns and CustomRuns, so there shouldn't be any surprises. 

In my mind, I imagined it might be possible to define a shared interface for all the `Run` objects, and remove most of the `if` statements in the code, but it probably gets complicated with resolution and in the places where the logic across resources diverges, so it's probably safer to go down this route.

/approve

**Comment by afrittoli** [comment]
The approvals are more than enough to merge. Minor comments on the comments can be addressed later.
/lgtm

**Comment by twoGiants** [comment]
> Thanks for this!
> 
> It looks good, it pretty much follows the patter of TaskRuns and CustomRuns, so there shouldn't be any surprises.
> 
> In my mind, I imagined it might be possible to define a shared interface for all the `Run` objects, and remove most of the `if` statements in the code, but it probably gets complicated with resolution and in the places where the logic across resources diverges, so it's probably safer to go down this route.
> 
> /approve

Thank you for mentioning this! I talked about it in the _Working Group_ meetings in May a few times. That is also the first came into my mind while deciding on how to implement it and it is for sure the better and more robust way for the future. But it needs **a lot** of refactorings. I started working on a POC refactoring in [this branch](https://github.com/twoGiants/tekton-pipeline/commit/12bdbde23d153a3dfafb5243bbbbf5763c00174a) in July but did not finish it, nor came I very far. I estimated at least two weeks of full time coding on it which is 80 hours, so we can safely double it and then triple or quadruple it when one starts fixing and moving the tests. 

I first thought to create an issue but then thought that nobody ever will touch this! :rofl: 

I wrote a few notes down while analyzing the code. It's a markdown file so I just copy it here :smile_cat: 

---

## Issue Header
Introduce polymorphism to `ResolvedPipelineTask`.
## Summary

## Proof of Concept

**WHAT YOU DID: 25.07**
- you created `pipelinerunresolutionV2.go` and started implementing the new data structure with three `ResolvedResource` types => you're focusing on `ResolvedChildPipelineRun` and the interface implementation => you decided to provide all the needed data during construction and put a few properties on `ResolvedChildPipelineRun`  and use it in the downstream methods
- create constructor for  `ResolvedChildPipelineRun`  => implement `ResolveResource` using `ResolvedChildPipelineRun` properties, then the same for `CustomRun` and `TaskRun`

**WHAT YOU DID: 26.07**
- implemented other resolved resources and added `pipelinerunstateV2.go` and `resolvePipelineStateV2()` in `pipelinerun.go`
- implemented `UpdateConditionFromVerificationResult` to simplify `resolvePipelineStateV2()` => now the power of polymorphism comes slowly to shine
- started looking into how to change those two functions in `ResolvePipelineTaskV2` => `ResolveResultRefs` and `ApplyTaskResults` so that you can finish this part

### Analysis Done
Check `ResolvedPipelineTask` and its descendants
- check all `ResolvedPipelineTask`  references and methods
	- **medium to hard but doable**, but a lot of moving around and decomposing
	- creating the child `PipelineRun/TaskRun/CustomRun` in `runNextSchedulableTask` is tricky => use strategy pattern with 3 creator algorithms and on `ResolvedPipelineTask` init pick the right one => later in `pipelinerun.go` it'll be used in `runNextSchedulableTask`
	- not all methods need refactor
- review `PipelineRunState` references => `[]*ResolvedPipelineTask`
	- **medium to hard but doable**, but a lot of moving around and decomposing
	- not all methods need refactor
	- first step is maybe (or not) to separate `PipelineRunState`, `PipelineRunFacts` and `ResolvedPipelineTask` which are all in one file => `pipelinerunstate.go`
	- `ConvertResultsMapToTaskRunResults` function should be a method on `rpt`
- review `PipelineRunState` methods => `[]*ResolvedPipelineTask`
	- **easy** => barely used
	- not all methods need refactor
- review `PipelineRunFacts` because it contains `PipelineRunState`
	- **easy** => you reviewed all the methods of `PipelineRunFacts` now review all references of `PipelineRunFacts`
	- only few methods need refactor
